Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,261 @@ Primary Objectives
+
+In RyuJIT, the concept of a type is very simplistic (which helps support the high throughput
+of the JIT). Rather than a symbol table to hold the properties of a type, RyuJIT primarily
+deals with types as simple values of an enumeration. When more detailed information is
+required about the structure of a type, we query the type system, across the JIT/EE interface.
+This is generally done only during the importer (translation from MSIL to the RyuJIT IR),
+during struct promotion analysis, and when determining how to pass or return struct values.
+As a result, struct types are generally treated as an opaque type
+(TYP_STRUCT) of unknown size and structure.
+
+In order to treat fully-enregisterable struct types as "first class" types in RyuJIT, we
+ created new types to represent vectors, in order for the JIT to support operations on them:
+* `TYP_SIMD8`, `TYP_SIMD12`, `TYP_SIMD16` and (where supported by the target) `TYP_SIMD32`.
+ - These types already exist, and represent some already-completed steps toward First Class Structs.
+
+ We propose to create the following additional types to be used where struct types of the given size are
+ passed and/or returned in registers:
+ * `TYP_STRUCT1`, `TYP_STRUCT2`, `TYP_STRUCT4`, `TYP_STRUCT8` (on 64-bit systems)
+
+For other struct types, for which the JIT has no type-specific optimizations, the following
+transformations need to be supported effectively:
+- Optimizations such as CSE and assertion propagation
+  - Depends on being able to discern when instances of these types are equivalent.
+- Passing and returning these values to and from methods
+  - Avoiding unnecessarily copying these values between registers or on the stack.
+  - Allowing promoted structs to be passed or returned without forcing them to become
+    ineligible for register allocation.
+
+Correct and effective code generation for structs requires that the JIT have ready access
+to information about the shape and size of the struct. This information is obtained from the VM
+over the JIT/EE interface. This includes:
+- Struct size
+- Number and type of fields, especially if there are GC references
+
+It has been proposed by @mikedn in
+[#21705 Pull struct type info out of GenTreeObj](https://github.com/dotnet/coreclr/pull/21705)
+that this kind of information could be kept in a `ClassLayout` cache, reducing the size impact on the `GenTree`
+nodes, as well as the number
+of JIT/EE interface queries. Reducing the impact on the size of the nodes would make it less
+costly to retain this information even on rvalues, thus enabling some
+optimization scenarios that are currently disabled for struct types.
+
+Current Representation of Struct Values
+---------------------------------------
+### Importer-Only Struct Values
+
+These struct-typed nodes are created by the importer, but transformed in morph, and so are not
+encountered by most phases of the JIT:
+* `GT_INDEX`: This is transformed to a `GT_IND`
```

Comment:
It may be good to mention somewhere the horrible complexity associated with array of structs element access - the `IND` must be preserved due to the array info map where it acts as a key. And then the whole element access tree tends to end up wrapped in `OBJ(ADDR(...))`. FWIW I'd love to get rid of that. At a minimum we should be able to change the `IND` to `OBJ` and avoid wrapping. We should also be able to remove the class handle from the array info map and instead used the one provided by `OBJ`.